File: HyeonHoLe2/3주차/PG_92334_신고_결과_받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
  
        Set<String> uniqueReports = new HashSet<>(Arrays.asList(report));
        
        Map<String, Set<String>> map = new HashMap<>(); 
        Map<String, Integer> map2 = new HashMap<>();    
        
        for(String r : uniqueReports){
            String[] parts = r.split(" ");
            String reporter = parts[0];
            String reported = parts[1];
            
           
            map2.put(reported, map2.getOrDefault(reported, 0) + 1);
            
         
            map.putIfAbsent(reporter, new HashSet<>());
            map.get(reporter).add(reported);
        }
        
        int[] answer = new int[id_list.length];
        
       
        for(int i = 0; i < id_list.length; i++){
            String user = id_list[i];
 
            Set<String> reportedList = map.getOrDefault(user, new HashSet<>());
            
            int mailCount = 0;
            for(String target : reportedList) {
          
                if(map2.getOrDefault(target, 0) >= k) {
                    mailCount++;
                }
            }
            answer[i] = mailCount;
        }
        
        return answer;
    }
}

File: KimJuha/3주차/PG_92334_신고결과받기.js
function solution(id_list, report, k) {
  const newReport = [...new Set(report)];

  const reportedCount = {};
  const reportedBy = {};

  for (let i = 0; i < id_list.length; i++) {
    reportedCount[id_list[i]] = 0;
    reportedBy[id_list[i]] = [];
  }

  for (let i = 0; i < newReport.length; i++) {
    const parts = newReport[i].split(" ");
    const reporter = parts[0];
    const reported = parts[1];

    reportedCount[reported]++;
    reportedBy[reporter].push(reported);
  }
  const result = [];

  for (let i = 0; i < id_list.length; i++) {
    const me = id_list[i];
    let mailCount = 0;

    for (let j = 0; j < reportedBy[me].length; j++) {
      const meReported = reportedBy[me][j];
      if (reportedCount[meReported] >= k) {
        mailCount++;
      }
    }
    result.push(mailCount);
  }
  return result;
}

// 시간복잡도는 O(n+m)
// 이중 for문처럼 보이지만 j는 me가 신고한 사람 수만큼만 반복됨
// -> newReport의 전체 갯수를 넘지 못함 (ReportedBy는 모두 newReport를 순회하면서 push 한 것들)
